File: 1/2021.04.08/Sort/java/MySortTest.java
package sort;

import org.junit.Assert;
import org.junit.Test;

import java.util.Arrays;
import java.util.Random;

public class MySortTest {

    private static final int LENGTH = 20000;
    private static final int BOUND = 10000000;
    private static final int TIMES = 10;


    @Test
    public void testBubbleSort() {
        int[] array = generateRandomArray();
        int[] array2 = Arrays.copyOf(array, array.length);
        Arrays.sort(array);
        MySort0408.bubbleSort(array2);
        Assert.assertArrayEquals(array, array2);
    }

    @Test
    public void testInsertSort() {
        int[] array = generateRandomArray();
        int[] array2 = Arrays.copyOf(array, array.length);
        Arrays.sort(array);
        MySort0408.insertSort(array2);
        Assert.assertArrayEquals(array, array2);
    }

    @Test
    public void testSelectionSort() {
        int[] array = generateRandomArray();
        int[] array2 = Arrays.copyOf(array, array.length);
        Arrays.sort(array);
        MySort0408.selectionSort(array2);
        Assert.assertArrayEquals(array, array2);
    }

    @Test
    public void testMergeSort() {
        int[] array = generateRandomArray();
        int[] array2 = Arrays.copyOf(array, array.length);
        Arrays.sort(array);
        array2 = MySort0408.mergeSort(array2);
        Assert.assertArrayEquals(array, array2);
    }

    @Test
    public void testQuickSort() {
        int[] array = generateRandomArray();
        int[] array2 = Arrays.copyOf(array, array.length);
        Arrays.sort(array);
        MySort0408.quickSort(array2);
        Assert.assertArrayEquals(array, array2);
    }

    @Test
    public void testBubblePerformance() {
        for (int i = 0; i < TIMES; i++) {
            int[] array = generateRandomArray();
            MySort0408.bubbleSort(array);
        }
    }

    @Test
    public void testMergePerformance() {
        for (int i = 0; i < TIMES; i++) {
            int[] array = generateRandomArray();
            MySort0408.mergeSort(array);
        }
    }

    @Test
    public void testQuickPerformance() {
        for (int i = 0; i < TIMES; i++) {
            int[] array = generateRandomArray();
            MySort0408.quickSort(array);
        }
    }

    @Test
    public void testSelectionPerformance() {
        for (int i = 0; i < TIMES; i++) {
            int[] array = generateRandomArray();
            MySort0408.selectionSort(array);
        }
    }

    @Test
    public void testInsertPerformance() {
        for (int i = 0; i < TIMES; i++) {
            int[] array = generateRandomArray();
            MySort0408.insertSort(array);
        }
    }

    private static int[] generateRandomArray() {
        int[] res = new int[LENGTH];
        Random random = new Random();
        for (int i = 0; i < res.length; i++) {
            res[i] = random.nextInt(BOUND);
        }
        return res;
    }
}

File: 1/2021.04.08/Sort/python/my_sort_test.py
import unittest
import random
from my_sort import bubble_sort, insert_sort, selection_sort, merge_sort, quick_sort

class TestCase(unittest.TestCase):
    def generate_array(self):
        res = []
        for _ in range(200):
            res.append(random.randint(0, 100000))
        return res
    
    def test_bubble_sort(self):
        arr = self.generate_array()
        arr_copy = arr[:]
        arr.sort()
        bubble_sort(arr_copy)
        self.assertEqual(arr, arr_copy)
    
    def test_insert_sort(self):
        arr = self.generate_array()
        arr_copy = arr[:]
        arr.sort()
        insert_sort(arr_copy)
        self.assertEqual(arr, arr_copy)

    def test_selection_sort(self):
        arr = self.generate_array()
        arr_copy = arr[:]
        arr.sort()
        selection_sort(arr_copy)
        self.assertEqual(arr, arr_copy)

    def test_merge_sort(self):
        arr = self.generate_array()
        arr_copy = arr[:]
        arr.sort()
        arr_copy = merge_sort(arr_copy)
        self.assertEqual(arr, arr_copy)

    def test_quick_sort(self):
        arr = self.generate_array()
        arr_copy = arr[:]
        arr.sort()
        quick_sort(arr_copy)
        self.assertEqual(arr, arr_copy)


if __name__ == '__main__':
    unittest.main()
